cpp/ql/src/experimental/Security/CWE/CWE-377/InsecureTemporaryFile.ql

```diff
@@ -0,0 +1,98 @@
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+/** Holds for a function `f` that has an argument at index `apos` used to read the file. */
+predicate numberArgumentRead(Function f, int apos) {
+  f.hasGlobalOrStdName("fgets") and apos = 2
+  or
+  f.hasGlobalOrStdName("fread") and apos = 3
+  or
+  f.hasGlobalOrStdName("read") and apos = 0
+  or
+  f.hasGlobalOrStdName("fscanf") and apos = 0
+}
+
+/** Holds for a function `f` that has an argument at index `apos` used to write to file */
+predicate numberArgumentWrite(Function f, int apos) {
+  f.hasGlobalOrStdName("fprintf") and apos = 0
+  or
+  f.hasGlobalOrStdName("fputs") and apos = 1
+  or
+  f.hasGlobalOrStdName("write") and apos = 0
+  or
+  f.hasGlobalOrStdName("fwrite") and apos = 3
+  or
+  f.hasGlobalOrStdName("fflush") and apos = 0
+}
+
+from FunctionCall fc, string msg
+where
+  (
+    fc.getTarget().hasGlobalOrStdName("tmpnam") or
+    fc.getTarget().hasGlobalOrStdName("tmpnam_s") or
+    fc.getTarget().hasGlobalOrStdName("tmpnam_r")
+  ) and
+  not exists(FunctionCall fctmp |
+    fctmp.getTarget().hasGlobalOrStdName("mktemp") or
+    fctmp.getTarget().hasGlobalOrStdName("mkstemp") or
+    fctmp.getTarget().hasGlobalOrStdName("mkstemps") or
+    fctmp.getTarget().hasGlobalOrStdName("mkdtemp")
+  ) and
+  msg =
+    "Finding the name of a file that does not exist does not mean that it will not be exist at the next operation."
+  or
+  (
+    fc.getTarget().hasGlobalOrStdName("fopen") or
+    fc.getTarget().hasGlobalOrStdName("open")
+  ) and
+  fc.getNumberOfArguments() = 2 and
+  exists(FunctionCall fctmp, int i |
+    numberArgumentWrite(fctmp.getTarget(), i) and
+    globalValueNumber(fc) = globalValueNumber(fctmp.getArgument(i))
+  ) and
+  not exists(FunctionCall fctmp, int i |
+    numberArgumentRead(fctmp.getTarget(), i) and
+    globalValueNumber(fc) = globalValueNumber(fctmp.getArgument(i))
+  ) and
+  exists(FunctionCall fctmp |
+    (
+      fctmp.getTarget().hasGlobalOrStdName("strcat") or
+      fctmp.getTarget().hasGlobalOrStdName("strcpy")
+    ) and
+    globalValueNumber(fc.getArgument(0)) = globalValueNumber(fctmp.getAnArgument())
+    or
+    fctmp.getTarget().hasGlobalOrStdName("getenv") and
+    globalValueNumber(fc.getArgument(0)) = globalValueNumber(fctmp)
+    or
+    (
+      fctmp.getTarget().hasGlobalOrStdName("asprintf") or
+      fctmp.getTarget().hasGlobalOrStdName("vasprintf") or
+      fctmp.getTarget().hasGlobalOrStdName("xasprintf") or
+      fctmp.getTarget().hasGlobalOrStdName("xvasprintf ")
+    ) and
+    exists(Variable vrtmp |
+      vrtmp = fc.getArgument(0).(VariableAccess).getTarget() and
+      vrtmp = fctmp.getArgument(0).(AddressOfExpr).getAddressable().(Variable) and
+      not vrtmp instanceof Field
+    )
+  ) and
+  not exists(FunctionCall fctmp |
+    fctmp.getTarget().hasGlobalOrStdName("umask") or
+    fctmp.getTarget().hasGlobalOrStdName("fchmod") or
+    fctmp.getTarget().hasGlobalOrStdName("chmod")
+  ) and
```

Comment:
I completely agree with your comments. I'm also totally fine with no changes to this piece of code. But note that the condition
```ql
not exists(FunctionCall fctmp |
  fctmp.getTarget().hasGlobalOrStdName("umask") or
  fctmp.getTarget().hasGlobalOrStdName("fchmod") or
  fctmp.getTarget().hasGlobalOrStdName("chmod")
)
```
will allow for false _negatives_. That is, you might fail to detect true vulnerabilities due to this condition being overly general.

Reply:
Good afternoon.
I suggest that you consider making changes to the `.qhelp` file so that the user does not hope for additional detection.

`Working with a file without the presence of calls in the code capable of evaluating its existence and rights, as well as working with predictable names, can be unsafe. Requires the attention of developers.`

also something like this can be inserted into the text of error messages.
if this solves the false negative problem, I will be glad.

Reply:
It doesn't really solve the problem, but it certainly makes the assumption of the query clearer. It's your choice whether you want to include the text or not.

Reply:
@MathiasVP please give me some more time to figure out what to do about this problem.

Reply:
Take all the time you want :)

Reply:
good day @MathiasVP.
see please, did you want such changes?

Reply:
Indeed. That was the kind of change I was hoping for, yes 👍